### PR TITLE
write multiple bytes of Char at once #27267

### LINKED ISSUE #27267: display() problem printing multi-bytes characters
```
julia> versioninfo()
Julia Version 0.7.0-DEV.5219
Commit b65712285a* (2018-05-25 14:37 UTC)
Platform Info:
  OS: Windows (i686-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-3630QM CPU @ 2.40GHz
  WORD_SIZE: 32
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, ivybridge)

julia> display("δ"); println("δ")
"��"
δ

julia> display("한"); println("한")
"���"
한

julia> display("aδ한a"); println("aδ한a")
"a�����a"
aδ한a

julia>
```

### PR DESCRIPTION
**Pull request by yeonsookimdev**
Fix #27267

`write()` function from 0.7.0-alpha:
```julia
function write(io::IO, c::Char) 
    # Let c == '한'
    # v == 0xed959c00
    #v = reinterpret(UInt32, c)
    
    # u = 0x009c95ed
    u = bswap(reinterpret(UInt32, c))
    
    # n |    1 |    2 |    3 |    4  
    #   | 0xed | 0x95 | 0x9c | 0x00
    n = 1
    while true
        # n == 1, write 0xed
        # n == 2, write 0x95
        # n == 3, write 0x9c
        write(io, u % UInt8)
        # return 3
        (u >>= 8) == 0 && return n
    n += 1
    end
end
```
```
julia> @time write(stdout, '한') == 3
���  0.000629 seconds (13 allocations: 400 bytes)
true

julia> @time write(stdout, '한') == 3
���  0.000445 seconds (10 allocations: 256 bytes)
true

julia> @time write(stdout, '한') == 3
���  0.000226 seconds (10 allocations: 256 bytes)
true

julia> @time write(stdout, '한') == 3
���  0.000305 seconds (10 allocations: 256 bytes)
true

julia> @time write(stdout, '한') == 3
���  0.000350 seconds (10 allocations: 256 bytes)
true

julia>
```

`write()` function from `alkorang/display`:
```julia
function write(io::IO, c::Char)
    # Let c == '한'
    # v == 0xed959c00
    v = reinterpret(UInt32, c)
    u = hton(v) # BIG-endian
    
    # p[0] | p[1] | p[2] | p[3]
    # 0xed | 0x95 | 0x9c | 0x00
    p = unsafe_convert(Ptr{Cvoid}, Ref{UInt32}(u))
    
    # Compare with v, not u
    if      v & 0x000000FF != 0
        unsafe_write(io, p, 4)
    elseif  v & 0x0000FFFF != 0
        # Write 0xed, 0x95, 0x9c at once, return 3
        unsafe_write(io, p, 3)
    elseif  v & 0x00FFFFFF != 0
        unsafe_write(io, p, 2)
    else
        unsafe_write(io, p, 1)
    end
end
```
```
julia> @time write(stdout, '한') == 3
한  0.000234 seconds (8 allocations: 320 bytes)
true

julia> @time write(stdout, '한') == 3
한  0.000253 seconds (5 allocations: 176 bytes)
true

julia> @time write(stdout, '한') == 3
한  0.000136 seconds (5 allocations: 176 bytes)
true

julia> @time write(stdout, '한') == 3
한  0.000183 seconds (5 allocations: 176 bytes)
true

julia> @time write(stdout, '한') == 3
한  0.000212 seconds (5 allocations: 176 bytes)
true

julia>
```

### CODE REVIEW

**Comment by alkorang** [comment]
> Any tests?

It passed `Base.runtests(["strings/io", "char", "iobuffer"])`.

```
julia> Base.runtests(["strings/io", "char", "iobuffer"]);
Test   (Worker) | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
strings/io  (2) |     4.76 |   0.14 |  3.0 |     326.25 |   239.07
char        (3) |     6.60 |   0.27 |  4.1 |     612.73 |   241.36
iobuffer    (2) |     5.00 |   0.13 |  2.7 |     661.03 |   260.16

Test Summary: |  Pass  Total
  Overall     | 14420  14420
    SUCCESS

julia>
```

**Comment by strickek** [comment]
Is speed for one byte Chars also better?

**Comment by alkorang** [comment]
> Is speed for one byte Chars also better?

```julia
function foo(io::IO, c::Char, n::Int)
    for _ in 1:5
        @time for _ in 1:1000
            write(io, c)
        end
    end
end

```
To make it more clear, I tried with `IOBuffer` after few warming-up.

0.7.0-alpha:
```
julia> GC.gc(); io = IOBuffer(); foo(io, 'a', 1000)
  0.000024 seconds (5 allocations: 2.172 KiB)
  0.000025 seconds (1 allocation: 2.125 KiB)
  0.000027 seconds (1 allocation: 2.000 KiB)
  0.000022 seconds
  0.000026 seconds (1 allocation: 4.000 KiB)

julia> GC.gc(); io = IOBuffer(); foo(io, 'δ', 1000)
  0.000041 seconds (6 allocations: 4.297 KiB)
  0.000042 seconds (1 allocation: 2.000 KiB)
  0.000050 seconds (1 allocation: 4.000 KiB)
  0.000043 seconds
  0.000049 seconds (1 allocation: 8.000 KiB)

julia> GC.gc(); io = IOBuffer(); foo(io, '한', 1000)
  0.000060 seconds (7 allocations: 6.297 KiB)
  0.000055 seconds (1 allocation: 4.000 KiB)
  0.000058 seconds (1 allocation: 8.000 KiB)
  0.000051 seconds
  0.000056 seconds

julia>
```

`alkorang/display`:
```
julia> GC.gc(); io = IOBuffer(); foo(io, 'a', 1000)
  0.000030 seconds (5 allocations: 2.172 KiB)
  0.000030 seconds (1 allocation: 2.125 KiB)
  0.000033 seconds (1 allocation: 2.000 KiB)
  0.000028 seconds
  0.000033 seconds (1 allocation: 4.000 KiB)

julia> GC.gc(); io = IOBuffer(); foo(io, 'δ', 1000)
  0.000034 seconds (6 allocations: 4.297 KiB)
  0.000034 seconds (1 allocation: 2.000 KiB)
  0.000033 seconds (1 allocation: 4.000 KiB)
  0.000035 seconds
  0.000043 seconds (1 allocation: 8.000 KiB)

julia> GC.gc(); io = IOBuffer(); foo(io, '한', 1000)
  0.000039 seconds (7 allocations: 6.297 KiB)
  0.000035 seconds (1 allocation: 4.000 KiB)
  0.000036 seconds (1 allocation: 8.000 KiB)
  0.000030 seconds
  0.000030 seconds

julia>
```


Is there any better way to measure them?

**Comment by nalimilan** [comment]
> Is there any better way to measure them?

Use `@btime` from BenchmarkTools.

File: base/io.jl [hunk 1]
```diff
@@ -563,12 +563,17 @@ function write(s::IO, a::SubArray{T,N,<:Array}) where {T,N}
...
 function write(io::IO, c::Char)
-    u = bswap(reinterpret(UInt32, c))
-    n = 1
-    while true
-        write(io, u % UInt8)
-        (u >>= 8) == 0 && return n
-        n += 1
+    v = reinterpret(UInt32, c)
+    u = hton(v) # BIG-endian
+    p = unsafe_convert(Ptr{Cvoid}, Ref{UInt32}(u))
```
Comment (JeffBezanson) [inline base/io.jl]: I believe this Ref needs a `GC.@preserve`.

Comment (alkorang) [reply base/io.jl]: Yes, it requires `GC.@preserve` for `Ref`.
```julia
function f0(c::Char)
    u = hton(reinterpret(UInt32, c))
    p = unsafe_convert(Ptr{Cvoid}, Ref{UInt32}(u))
    show(unsafe_wrap(Array{UInt8,1}, Ptr{UInt8}(p), (4,)))
end

function f1(c::Char)
    u = hton(reinterpret(UInt32, c))
    GC.@preserve u begin
        p = unsafe_convert(Ptr{Cvoid}, Ref{UInt32}(u))
        show(unsafe_wrap(Array{UInt8,1}, Ptr{UInt8}(p), (4,)))
    end
end

function f2(c::Char)
    u = Ref{UInt32}(hton(reinterpret(UInt32, c)))
    GC.@preserve u begin
        p = unsafe_convert(Ptr{Cvoid}, u)
        show(unsafe_wrap(Array{UInt8,1}, Ptr{UInt8}(p), (4,)))
    end
end
```

`UInt8[0xed, 0x95, 0x9c, 0x00]` must be printed:
```
julia> f0('한')
UInt8[0x40, 0x27, 0x3e, 0x3c]
julia> f0('한')
UInt8[0x01, 0x00, 0x00, 0x00]
julia>

julia> f1('한')
UInt8[0x20, 0x1c, 0x3e, 0x3c]
julia> f1('한')
UInt8[0x01, 0x00, 0x00, 0x00]
julia>

julia> f2('한')
UInt8[0xed, 0x95, 0x9c, 0x00]
julia> f2('한')
UInt8[0xed, 0x95, 0x9c, 0x00]
julia>
```

Comment (vtjnash) [reply base/io.jl]: Right – if `@time` isn't showing the extra allocation of the `Ref` object, this is broken (as you are demonstrating above)

**Comment by appleparan** [comment]
> It passed Base.runtests(["strings/io", "char", "iobuffer"]).

I mean for your issue. e.g. '한'. I think you should create tests for this issue.

**Comment by alkorang** [comment]
> I mean for your issue. e.g. '한'. I think you should create tests for this issue.

https://github.com/JuliaLang/julia/blob/5583ae42d5b541c1fa96f68c75f6e4bbff5c4740/test/char.jl#L213-L239

Here are some tests for multibyte cases. `test_overlong` calls `sprint(show,  c)` which eventually calls `write(io::IO, c::Char)`. If `write` did not work correctly, it cannot pass these tests.

File: base/io.jl [hunk 2]
```diff
@@ -563,12 +563,19 @@ function write(s::IO, a::SubArray{T,N,<:Array}) where {T,N}
...
 
 function write(io::IO, c::Char)
-    u = bswap(reinterpret(UInt32, c))
-    n = 1
-    while true
-        write(io, u % UInt8)
-        (u >>= 8) == 0 && return n
-        n += 1
+    v = reinterpret(UInt32, c)
+    u = Ref{UInt32}(hton(v)) # BIG-endian
```
Comment (StefanKarpinski) [inline base/io.jl]: This should be an unconditional `bswap` since UTF-8 is independent of the endianess of the system.

File: base/io.jl [hunk 3]
```diff
@@ -563,12 +563,19 @@ function write(s::IO, a::SubArray{T,N,<:Array}) where {T,N}
...
+        p = unsafe_convert(Ptr{Cvoid}, u)
+        if      v & 0x000000FF != 0
+            unsafe_write(io, p, 4)
+        elseif  v & 0x0000FFFF != 0
+            unsafe_write(io, p, 3)
+        elseif  v & 0x00FFFFFF != 0
+            unsafe_write(io, p, 2)
+        else
+            unsafe_write(io, p, 1)
+        end
```
Comment (StefanKarpinski) [inline base/io.jl]: You can replace this with something along the lines of `unsafe_write(io, p, (leading_zeros(u) >>> 3) + 1)`.

File: base/io.jl (hunk 2 above)
Comment (alkorang) [reply base/io.jl]: > UTF-8 is independent of the endianess of the system.

Yes, it is true, but `UInt32` isn't.

File: base/io.jl (hunk 3 above)
Comment (alkorang) [reply base/io.jl]: It's a good idea!

**Review by StefanKarpinski — APPROVED** [review]
Looks good to me. Shame about AppVeyor timing out. Perhaps we should wait until that's sorted out and rebase since testing Windows seems fairly key here.

**Comment by alkorang** [comment]
> Perhaps we should wait until that's sorted out and rebase since testing Windows seems fairly key here.

@StefanKarpinski @JeffBezanson
Some other tests fail on Windows. I first thought it will be okay if I rebase, but no.

One of them is from `cmdlineargs.jl`.
https://github.com/JuliaLang/julia/blob/b28496a7ae8cc0d7c3d175d2fba8757f2978d788/test/cmdlineargs.jl#L544-L545

Error message from 64bit AppVeyor
```
Error in testset cmdlineargs:
Test Failed at C:\projects\julia\julia-df4bb5401f\share\julia\test\cmdlineargs.jl:545
  Expression: read(pipeline(exename, stdin=infile), String) == "(1, 5)\n"
   Evaluated: "\x001, 5)\n" == "(1, 5)\n"
```
Error message from 32bit AppVeyor
```
Test Failed at C:\projects\julia\julia-df4bb5401f\share\julia\test\cmdlineargs.jl:545
  Expression: read(pipeline(exename, stdin=infile), String) == "(1, 5)\n"
   Evaluated: "\xcc1, 5)\n" == "(1, 5)\n"
ERROR: LoadError: Test run finished with errors
```

I guess the problem is from garbage collected values, because it keep changes like `\x00` or `\xd0` on my computer. But I don't understand. I protect `Ref` with `GC.@preserve` and it should never collect it. 

These failed tests are spawning a process and read from or write to `stdin`/`stdout` using `@cmd`, `spawn`, or `pipeline`.

**Comment by alkorang** [comment]
@vtjnash
I think this pr won't be available until finding out the reason of broken bytes from `unsafe_write`. Fixing libuv to use buffering looks more general and nicer.

**Comment by alkorang** [comment]
#28314 Fixes the problem.

**Pull request closed without merging**